### PR TITLE
Upgrade various libraries and dependencies to latest versions.

### PR DESCRIPTION
**Pull request by myohn-paypal**
```
Upgrade various libraries and dependencies to latest versions.

Upgrade various libraries, in particular:
spring-boot 1.4 -> 2.5
spring-framework 4.3 -> 5.3

Fixes issue #374
```

### CODE REVIEW

File: butterfly-test/src/main/resources/log4j2.xml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	<Appenders>
 		<Console name="CONSOLE" target="SYSTEM_OUT" follow="true">
 			<PatternLayout>
-				<Pattern>[%d{HH:mm:ss.SSS}] [%highlight(%level)] %msg%n</Pattern>
+				<Pattern>[%d{HH:mm:ss.SSS}] [%highlight{%level}] %msg%n</Pattern>
```
Comment (fabiocarvalho777) [inline butterfly-test/src/main/resources/log4j2.xml:6]: Why is this change necessary? This works as is.

File: butterfly-cli/src/test/resources/helpOut.txt [hunk 2]
```diff
@@ -20,7 +20,7 @@ Option                           Description
                                    application is placed under a new folder    
                                    whose name is same as original folder, plus 
                                    "-transformed-yyyyMMddHHmmssSSS" suffix     
--p <inline properties>           Transformation specific properties, used to   
+-p <String: inline properties>   Transformation specific properties, used to
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/test/resources/helpOut.txt:23]: What is the reason for this change? This is CLI help information. How adding `String` here would help an user when typing the command in the CLI? By the way, String is a Java specific term but the CLI user doesn't even have to know that.

File: butterfly-cli/build.gradle [hunk 3]
```diff
@@ -12,20 +12,45 @@ dependencies {
...
         exclude(module: 'guava')
     }
     testCompile lib.mockito_all,
-            lib.powermock_module_testng,
-            lib.powermock_api_mockito
+                lib.hamcrest,
+                lib.javassist,
+                lib.objenesis
+    testCompile(lib.powermock_api_mockito) {
+        // favor version pulled in by lib.hamcrest
```
Comment (fabiocarvalho777) [inline butterfly-cli/build.gradle:37]: These exclusions shouldn't be necessary. Gradle is supposed to take care of it. Also, not always the newer version is backward compatible, which may cause runtime issues.

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: butterfly-cli/src/test/resources/helpOut.txt (hunk 2 above)
Comment (myohn-paypal) [reply butterfly-cli/src/test/resources/helpOut.txt:23]: newer version of jopt insists on showing the type. there does not seem to be an easy way to disable that.

see:
[new version](https://github.com/jopt-simple/jopt-simple/blob/jopt-simple-5.0.4/src/main/java/joptsimple/BuiltinHelpFormatter.java#L436-L443)
[current version](https://github.com/jopt-simple/jopt-simple/blob/jopt-simple-5.0.2/src/main/java/joptsimple/BuiltinHelpFormatter.java#L432-L439)

File: butterfly-cli/build.gradle (hunk 3 above)
Comment (myohn-paypal) [reply butterfly-cli/build.gradle:37]: This was an attempt at fixing conflicting dependency versions. Gradle might "take care of it" but the fact remains that the app still pulls in conflicting dependencies. Without this exclusion, the app would pull in:

```
+--- org.powermock:powermock-api-mockito:1.6.5
 |    +--- org.mockito:mockito-core:1.10.19
 |    |    \--- org.hamcrest:hamcrest-core:1.1               <---
 |    +--- org.hamcrest:hamcrest-core:1.3
 |    \--- org.powermock:powermock-api-mockito-common:1.6.5
 |         +--- org.hamcrest:hamcrest-core:1.3               <---
```

... meaning the app has some deps that want hamcrest 1.1 and some deps that want hamcrest 1.3. This exclusion ensures a deliberate choice for which version of hamcrest the app should use.

&nbsp;
Also, I say attempt because I gave up half-way through. Although I fixed conflicts for compile-time deps, conflicts still remain for test-only deps, which require more effort to fix (and are arguably less important)

Comment (myohn-paypal) [reply butterfly-cli/build.gradle:37]: I think there should be another pass or pull request to upgrade the test deps. There we could remove the manual exclusions like these.

File: butterfly-cli/src/test/resources/helpOut.txt (hunk 2 above)
Comment (fabiocarvalho777) [reply butterfly-cli/src/test/resources/helpOut.txt:23]: Ok, thanks

File: butterfly-test/src/main/resources/log4j2.xml (hunk 1 above)
Comment (fabiocarvalho777) [reply butterfly-test/src/main/resources/log4j2.xml:6]: But indeed, that link show the usage of curly braces instead of parenthesis. Did you test this change? Did the loglevel get colored?

File: butterfly-cli/build.gradle (hunk 3 above)
Comment (fabiocarvalho777) [reply butterfly-cli/build.gradle:37]: I see. Thanks for clarifying Myo!

File: butterfly-test/src/main/resources/log4j2.xml (hunk 1 above)
Comment (fabiocarvalho777) [reply butterfly-test/src/main/resources/log4j2.xml:6]: Cool, thanks

**Review by fabiocarvalho777 — APPROVED** [review]

File: butterfly-test/src/main/resources/log4j2.xml (hunk 1 above)
Comment (myohn-paypal) [reply butterfly-test/src/main/resources/log4j2.xml:6]: Strange though that highlights worked for you. hmm
